rfd/9999-passwordless.md

```diff
@@ -0,0 +1,501 @@
+
+No further request or response message changes are necessary, but it's important
+to note that the user handle in [CredentialAssertionResponse](
+https://github.com/gravitational/teleport/blob/f423f7fedc088b97cb666c13dcdcf54bd289b1bf/api/types/webauthn/webauthn.proto#L81)
+is required for passwordless logins.
+
+Finally, un-authenticated login endpoints are now subject to IP-based rate
+limiting, as it seems wise to further safeguard system resources from anonymous
+access.
+
+### Registration
+
+Registration requires us to know whether the user intends use to the device for
+passwordless, so we can request a resident key and ask for the appropriate user
+verification requirements.
+
+Requesting resident keys is a delicate decision, as most FIDO2 devices have
+limited capacity for resident keys (eg, [25 keys for Yubico FIDO2](
+https://support.yubico.com/hc/en-us/articles/360013790319-How-many-credentials-can-I-program-on-my-YubiKey-)).
+Users may want to pick and choose depending on how many accounts they have.
+Platform authenticators may provide limits large enough to make it a non-issue,
+so defaulting to passwordless/resident keys on them may be an easier decision.
+
+<!--
+TODO(codingllama): Sources needed for platform resident key limits.
+-->
+
+The backend makes no assumptions and lets the clients/UI decide, as they are
+better equipped to do so.
+
+The following changes are made to registration messages:
+
+<!--
+CreateRegisterChallenge: https://github.com/gravitational/teleport/blob/f423f7fedc088b97cb666c13dcdcf54bd289b1bf/api/client/proto/authservice.proto#L1354
+
+AddMFADeviceRequestInit: https://github.com/gravitational/teleport/blob/f423f7fedc088b97cb666c13dcdcf54bd289b1bf/api/client/proto/authservice.proto#L878
+-->
+
+```proto
+package proto; // api/client/proto
+
+enum DeviceUsage {
+  DEVICE_USAGE_UNSPECIFIED = 0;
+
+  // Device intended for MFA use, but not for passwordless.
+  // Allows both FIDO and FIDO2 devices.
+  // Resident keys not required.
+  DEVICE_USAGE_MFA = 1;
+
+  // Device intended for both MFA and passwordless.
+  // Requires a FIDO2 device and a resident key slot.
+  DEVICE_USAGE_PASSWORDLESS = 2;
+}
+
+message CreateRegisterChallengeRequest {
+  // (existing fields omitted)
+
+  // Requested usage for the device.
+  // Defaults to DEVICE_USAGE_MFA.
+  DeviceUsage device_usage = 3;
+}
+
+message AddMFADeviceRequestInit {
+  // (existing fields omitted)
+
+  // Requested usage for the device.
+  // Defaults to DEVICE_USAGE_MFA.
+  DeviceUsage device_usage = 4;
+}
+```
+
+#### Device restrictions
+
+Passwordless device restrictions can be attained through [attestation](
+https://github.com/gravitational/teleport/blob/f423f7fedc088b97cb666c13dcdcf54bd289b1bf/rfd/0040-webauthn-support.md#attestation).
+
+Teleport may choose to ship with canned set of allowed passwordless roots. An
+example of said roots could include:
+
+* [Apple WebAuthn Root CA](https://www.apple.com/certificateauthority/private/)
```

Comment:
I'd be leaning on canned retrieval here, but I'll make a note in the design. Do you have any specific checks in mind, in any case?

Reply:
For the FIDO Metadata Service:
* Follow the [metadata TOC processing rules](https://fidoalliance.org/specs/fido-v2.0-rd-20180702/fido-metadata-service-v2.0-rd-20180702.html#metadata-toc-object-processing-rules) and verify the signature of the Metadata TOC object using the TOC signing certificate chain
* Check whether the status report of the authenticator model has changed compared to the cached entry (e.g. an authenticator changed from the status `REVOKED` to `FIDO_CERTIFIED` unexpectedly)
* Cross-reference the TOC returned with third party metadata (the relying party, Teleport, should obtain metadata directly from authenticator vendors or other trusted sources and cross-checks it)
* Raise a log warning if it changes from the latest cached version

For the Apple attestation statement it could be harder, but you could try to pin the TLS certificate for the endpoint if it's observed to be stable.